Add Brave Wallet

kylehickinson opened the pull request:
This is a cumulative PR that includes many issues found in the 1.33 milestone/Brave Wallet project

Note: Merge with Merge Commit, not rebase/squash (best to keep git history for this large of a change)

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

kylehickinson committed:
Move Brave Wallet work into brave-ios

kylehickinson committed:
Cleanup unused images, manual AutoLayout and immediate closure execs

kylehickinson committed:
Disable landscape orientation on wallet

kylehickinson committed:
Fix wallet dismiss from onboarding & unlock. Fix settings icon tints

kylehickinson committed:
Cleanup in portfolio & backup banner

kylehickinson committed:
Update biometrics enable popup view

kylehickinson committed:
Replace "Live" with "1H" in date range pickers

kylehickinson committed:
Basic settings view for resetting wallet

kylehickinson committed:
Add alternate gradient02 gradient and use it for buy/send/swap button

kylehickinson committed:
Show onboarding again immediately when resetting from within wallet

kylehickinson committed:
Add ability to import legacy brave wallet when phrase is 24 words

kylehickinson committed:
Switch usages of UIKitScrollView to ScrollView

Keyboard avoidance is no longer an issue since the deployment target for wallet is now 14+

kylehickinson committed:
Fix non-Text Section headers display

This resets the font, foreground color, list row insets and text case for Section headers that are not used to display Text

kylehickinson committed:
Add biometrics unlock support

kylehickinson committed:
Fix iOS 15 bugs related to root StackNavigationViewStyle and reset

kylehickinson committed:
Merge development into brave-wallet

kylehickinson committed:
Fix toolbar button colors on iOS 15

kylehickinson committed:
SwiftUI extensions for BraveWallet mojo structs

kylehickinson committed:
Fixup codesigning

kylehickinson committed:
Nightly bump

kylehickinson committed:
Move iOS deployment target to iOS 14

kylehickinson committed:
Remove iOS 14 availability annotations, add reset list style modifier

WIP on crypto swap UI

kylehickinson committed:
Better account name prompt and creation logic

The name of the account is now optional and will fallback automatically to "Account {N+1}" where N is the number of accounts in that category you already have.

kylehickinson committed:
Add UI to export primary and secondary account private keys

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Remove target-specific deployment targets

kylehickinson committed:
WIP: Wallet Store

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Fix #4170: Refactor fetching brave-core APIs for sync related classes (#4220)

Also applies other needed build fixes for 1.32.x

kylehickinson committed:
Nightly bump

kylehickinson committed:
Fix #4261: Handle identifiable conformance for ETH and add swap address

ETH now uses an empty contract address

kylehickinson committed:
Fix #4265: Reset biometric unlock when restoring a wallet successfully

kylehickinson committed:
Fix #4268: Finish asset search UI & integration

kylehickinson committed:
Nightly bump to 1.32.44

kylehickinson committed:
Fix #4269: Add support for editing visible assets

kylehickinson committed:
Merge development into brave-wallet

kylehickinson committed:
Fix #4260: Handle keyring auto-lock based on user interaction

kylehickinson committed:
Fix #4259: Replace hard-coded user facing strings with NSLocalizedString (#4310)

kylehickinson committed:
Nightly bump to 1.32.63

kylehickinson committed:
Fix #4302: Add setting to adjust auto-lock interval

kylehickinson committed:
Fix release builds

kylehickinson committed:
Fix #4311: Always display ETH as an option for user assets

kylehickinson committed:
Fix #4315: Add button to create/import accounts from the account picker

kylehickinson committed:
Fix copy on backup & account private key screens

Also disables `LOCALIZED_STRING_SWIFTUI_SUPPORT` which makes Xcode export string literals used in SwiftUI Text initializers since we will never use that.

kylehickinson committed:
Add more button to unlocked wallet and refactored CryptoPagesView

Also fixes the safe area on unlock & onboarding

kylehickinson committed:
Hookup account picker to selected accounts

nuo-xu committed:
Fix #4272: Add Buy Crypto UI (#4327)

kylehickinson committed:
Fix #4279: Add support for displaying icons for token registry assets (#4326)

kylehickinson committed:
Refactor buy/send/swap actions into SwiftUI presentation context

nuo-xu committed:
Merge pull request #4328 from brave/buy-send-swap-destination-refactor

Refactor buy/send/swap actions into SwiftUI presentation context

kylehickinson committed:
Ref #4272: Add support for opening wallet URLs (#4345)

kylehickinson committed:
Merge origin/development into brave-wallet

nuo-xu committed:
Ref #4272: Wallet Buy Screen enhancement  (#4344)

kylehickinson committed:
Fix potential memory leak by weakifying BraveWalletDelegate

kylehickinson committed:
Fix unit test build error after bad merge conflict

kylehickinson committed:
Update Brave Wallet strings (#4349)

kylehickinson committed:
Bump BraveCore build to Beta 1.32.73 (#4350)

kylehickinson committed:
Immediately focus the password field on unlock screen (#4351)

Also add a workaround for using biometrics unlock in the simulator

nuo-xu committed:
Fix: #4371 Test network UI in Buy Screen (#4392)

kylehickinson committed:
Bump BraveCore build to beta 1.32.79

- Includes fixes for crash when creating wallet (localized strings now bundled)
- Includes infura ID for eth rpc calls

kylehickinson committed:
Fix #4262: Portfolio balance & historical balances (#4410)

nuo-xu committed:
Fix: #4273 Built Send Crypto Screen  (#4380)

kylehickinson committed:
Fix #4418: Update Brave Wallet icon

kylehickinson committed:
Fix #4417: Update password validation requirements

kylehickinson committed:
Add wallet settings access to main settings

kylehickinson committed:
Fix portfolio assets showing incorrect balance

kylehickinson committed:
Add empty balance history state to portfolio

kylehickinson committed:
Fix #4389: Update backup banner background color (#4427)

kylehickinson committed:
Add support for converting decimal to hex wei string (#4428)

nuo-xu committed:
Fix: #4422 Send Screen with Unapproved Transaction API (#4446)

kylehickinson committed:
Add convenience method to fetch decimal balance for any token (#4456)

kylehickinson committed:
Fix #4270: Asset detail price & price history integration (#4458)

nuo-xu committed:
Finishes Swap UI, adds partial swap integration & disables access to swap on testnets (#4472)

kylehickinson committed:
Fix #4444: Transaction confirmation & edit gas UI (#4481)

kylehickinson committed:
Fix #4320: Add numbers next to backup phrase during onboarding (#4482)

kylehickinson committed:
No Bug: Update copy on setup crypto view (#4489)

kylehickinson committed:
Fix #4420: Add support for setting up biometric unlock after restore (#4484)

Also adds a explicit button to fill password using biometrics in the unlock wallet view

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Fix #4266: Update onboarding/unlock wallet illustrations (#4488)

nuo-xu committed:
Fix #4271:  Integrate Swap Price Quote and related user cases. (#4499)

nuo-xu committed:
Fix: #4503 Added more error handling in swap (#4507)

kylehickinson committed:
Bump BraveCore to 1.33.75 (#4512)

nuo-xu committed:
Fix: #4493 Hardcode swap assets in Ropsten (#4511)

kylehickinson committed:
Fix #4459, Fix #4274: Add transactions list to asset details & accounts (#4514)

kylehickinson committed:
Fix CryptoPagesView sheet presentation on iOS 14.0-14.4 (#4519)

kylehickinson committed:
Fix #4487: Fetch portfolio details if wallet is already unlocked (#4531)

kylehickinson committed:
Fix colour for unapproved transactions and update localization comment (#4532)

nuo-xu committed:
Fix: 4294 support wallet add custom asset (#4538)

nuo-xu committed:
Add support for making EIP-1559 send & swap transactions on supported chains (#4545)

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Fix #4556: Remove localhost option from network selection list (#4557)

nuo-xu committed:
Fix #4554: Filter out erc721 tokens in token selection inside asset search, send and swap  (#4558)

nuo-xu committed:
Fixed swap button gets reset after an error occurs during adding unapproved tx. Fixed the crash caused by gas estimation hex conversion. (#4564)

iccub reviewed (commented):
nitpick review, did not find anything concerning really

Before final review I would consider running through this code to check for any TODO and FIXME comments,
and check for autolocalized `Text`s. However I saw a flag saying to not auto translate them so not a big deal

Only code architecture comment I have is about hardcoded usd values for fiat, if we plan to add more currencies in the future we could prepare for it better. I do not think it's a blocker in any way now.
Just sometimes you have to watch out when working in currencies, for example sometimes depending on country and region, currency symbol is placed before or after amount.
This makes certain placeholder or fallback values to look incorrectly

kylehickinson committed:
Cleanup wallet, fix missed localizations, remove unused todos/files (#4569)

kylehickinson committed:
Fix #4568: Hide Brave Wallet access in private mode, and fix menu copy (#4573)

kylehickinson committed:
Fix #4445: Integrate transaction confirmation UI with tx controller (#4567)

nuo-xu committed:
Wallet: Added missing localized strings. (#4566)

jamesmudgett committed:
Update WalletStrings.swift

Updated CTA, confirming other valid strings.

jamesmudgett committed:
Update WalletStrings.swift

Update feature title, and buy description for buy/send/swap widget.

kylehickinson committed:
Fix #4461: Add iOS 14 specific bug fixes (#4574)

kylehickinson committed:
Fix majority of SonarCloud code-smells

kylehickinson committed:
Fix #4579: Add note about using 0x to swap UI (#4581)

nuo-xu committed:
Fix #4413: Reset all the stores except KeyringStore after user resets wallet (#4585)

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Fix #4303: Improve VoiceOver accessibility (#4580)

kylehickinson committed:
Fix #4591: Show wallet button in menu while private browsing (#4592)

kylehickinson committed:
Fix #4595: Ensure we update keyring state when app foregrounds (#4596)

Also harden logic around requesting keychain data on unlock screen

kylehickinson committed:
Fix #4597: Fix crash when editing a custom gas fee (#4598)

This is caused by an iOS 14 bug where an animation on the list is causing a deferred addition of the "custom" section which is getting set on `onAppear`. By switching to use an animation on the `Binding` of `gasFeeKind` we only animate explicitly on changing `gasFeeKind` via the `Picker`

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Use new shared password validation from BraveCore (#4606)

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Update copy for new password requirements (#4614)

kylehickinson committed:
Fix asset details losing state after changing network or adding accounts (#4616)

kylehickinson committed:
Fix #4617, Fix #4618: Add CoinGecko and Brave swap fee disclosures (#4621)

kylehickinson committed:
No Bug: Change password fields text content types for create/restore wallet (#4622)

kylehickinson committed:
Hide private key when backgrounding and embed screen in ScrollView (#4612)

Also fixes the preview account address to match the test keyring controller address and small change to swap code

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Fix #4613: Make private key and recovery phrase copy to clipboard secure (#4624)

kylehickinson committed:
Merge origin/development into brave-wallet

iccub committed:
No bug: Wallet duplicate string keys (#4630)

iccub reviewed (approved):
Amazing work and showcase of SwiftUI possibilities :) 
my part of the review looks good

nuo-xu committed:
Fix #4504: Destroy store when associated view is gone. (#4615)

kylehickinson committed:
Fix #4633: Properly dismiss nested wallet sheets when the wallet locks (#4634)

nuo-xu committed:
Fix #4620: Disable action button and add a spinner on top while send/swap unapproved tx is under construction.  (#4639)

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Add error code when failing to setup biometrics in a non-public build (#4647)

kylehickinson committed:
Fix crash when balance calls fail; Fix small SwiftUI bug (#4642)

kylehickinson committed:
Fix #4643: Don't capture private data while screen recording (#4644)

Also displays an alert when you screenshot on screens displaying private data

nuo-xu committed:
Fix #4641: Add send address validation in send screen. (#4646)

kylehickinson committed:
Merge origin/development into brave-wallet

kylehickinson committed:
Fix possible crash when opening wallet (#4661)

kylehickinson committed:
Fix #4659: Add beta label to wallet in menu and add subtitle in NTP (#4658)

kylehickinson committed:
Fix adding custom assets with different decimals

kylehickinson committed:
Fix sending custom assets and showing wrong value for ERC20 transactions

kylehickinson committed:
Fix transaction details for ERC20 transfers

kylehickinson committed:
Fix #4666: Fix crash if chain is changed when price request is in flight

kylehickinson committed:
Fix #4669: Fix user interaction notifier removal on wallet deinit (#4670)

kylehickinson committed:
Fix #4667: Wait for send/swap to fully disappear before showing tx confirmations (#4671)

kylehickinson committed:
Merge origin/development into brave-wallet
